core/src/main/scala/quasar/plugin/postgres/datasource/PostgresDatasource.scala

```diff
@@ -0,0 +1,224 @@
+            .map(p => if (p) Some(Stream.empty.covaryAll[F, (ResourceName, RPT.Physical)]) else None)
+            .transact(xa)
+
+        case Some(Left(schema)) =>
+          val l = Some(schema).filterNot(containsWildcard).map(Left(_))
+
+          val paths =
+            tables(l)
+              .filter(_.schemaName === schema)
+              .map(m => (ResourceName(m.tableName), RPT.leafResource))
+              .pull.peek1
+              .flatMap(t => Pull.output1(t.map(_._2)))
+              .stream
+
+          val back = for {
+            c <- xa.connect(xa.kernel)
+            opt <- paths.translate(runCIO(c)).compile.resource.lastOrError
+          } yield opt.map(_.translate(runCIO(c)))
+
+          back.allocated flatMap {
+            case (o @ None, release) => release.as(o)
+            case (Some(s), release) => s.onFinalize(release).some.pure[F]
+          }
+
+        case None => (None: Option[Stream[F, (ResourceName, RPT.Physical)]]).pure[F]
+      }
+
+  ////
+
+  private type Loc = Option[Either[Schema, (Schema, Table)]]
+
+  // Characters considered as pattern placeholders in
+  // `DatabaseMetaData#getTables`
+  private val Wildcards: Set[Char] = Set('_', '%')
+
+  // We use `tables` here to constrain discovered schemas
+  // to be only those containing visible tables.
+  private def allSchemas: Stream[ConnectionIO, String] =
+    tables(None)
+      .scan((Set.empty[String], None: Option[String])) {
+        case ((seen, _), TableMeta(schema, _)) =>
+          if (seen(schema))
+            (seen, None)
+          else
+            (seen + schema, Some(schema))
+      }
+      .map(_._2)
+      .unNone
+
+  /** Returns a `Loc` representing the most specific tables selector.
+    *
+    * To be a tables selector, an `Ident` must not contain any of the wildcard characters.
+    */
+  private def locSelector(schema: Schema, table: Table): Loc =
+    (containsWildcard(schema), containsWildcard(table)) match {
+      case (true, _) => None
+      case (false, true) => Some(Left(schema))
+      case (false, false) => Some(Right((schema, table)))
+    }
+
+  private def containsWildcard(s: String): Boolean =
+    s.exists(Wildcards)
+
+  /** Returns a quoted and escaped version of `ident`. */
+  private def hygienicIdent(ident: Ident): Ident =
+    s""""${ident.replace("\"", "\"\"")}""""
+
+  private def pathToLoc(rp: ResourcePath): Loc =
+    Some(rp) collect {
+      case schema /: ResourcePath.Root => Left(schema)
+      case schema /: table /: ResourcePath.Root => Right((schema, table))
+    }
+
+  private def runCIO(c: java.sql.Connection): ConnectionIO ~> F =
+    λ[ConnectionIO ~> F](_.foldMap(xa.interpret).run(c))
+
+  private def tableAsJsonBytes(schema: Schema, table: Table): Stream[ConnectionIO, Byte] = {
+    val schemaFr = Fragment.const0(hygienicIdent(schema))
+    val tableFr = Fragment.const(hygienicIdent(table))
+    val sql = fr"SELECT to_json(t) FROM" ++ schemaFr ++ fr0"." ++ tableFr ++ fr0"AS t"
```

Comment:
What are `fr` and `fr0`? Where do they come from?

Reply:
They're sql fragment constructors from `doobie` the `0` version omits the default trailing space.